source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,163 @@
+                                  socklen_t optlen) override;
+  Api::SysCallIntResult getOption(int level, int optname, void* optval, socklen_t* optlen) override;
+  Api::SysCallIntResult setBlocking(bool blocking) override;
+  absl::optional<int> domain() override;
+  Address::InstanceConstSharedPtr localAddress() override;
+  Address::InstanceConstSharedPtr peerAddress() override;
+  Event::FileEventPtr createFileEvent(Event::Dispatcher& dispatcher, Event::FileReadyCb cb,
+                                      Event::FileTriggerType trigger, uint32_t events) override;
+  Api::SysCallIntResult shutdown(int how) override;
+  absl::optional<std::chrono::milliseconds> lastRoundTripTime() override { return absl::nullopt; }
+
+  Buffer::WatermarkBuffer& getBufferForTest() { return owned_buffer_; }
+
+  void scheduleNextEvent() {
+    // It's possible there is no pending file event so as no io_callback.
+    if (io_callback_) {
+      ENVOY_LOG(trace, "Schedule IO callback on {}", static_cast<void*>(this));
+      io_callback_->scheduleCallbackNextIteration();
+    }
+  }
+
+  void setWritablePeer(WritablePeer* writable_peer) {
+    // Swapping writable peer is undefined behavior.
+    ASSERT(!writable_peer_);
+    ASSERT(!write_shutdown_);
+    writable_peer_ = writable_peer;
+  }
+
+  // WritablePeer
+  void setWriteEnd() override { read_end_stream_ = true; }
+  bool isWriteEndSet() override { return read_end_stream_; }
+  void maybeSetNewData() override {
+    ENVOY_LOG(trace, "{} on socket {}", __FUNCTION__, static_cast<void*>(this));
+    scheduleNextEvent();
+  }
+  void onPeerDestroy() override {
+    writable_peer_ = nullptr;
+    write_shutdown_ = true;
+  }
+  void onPeerBufferWritable() override { scheduleNextEvent(); }
+  bool isWritable() const override { return !isOverHighWatermark(); }
+  Buffer::Instance* getWriteBuffer() override { return &owned_buffer_; }
+
+  // ReadableSource
+  bool isPeerShutDownWrite() const override { return read_end_stream_; }
+  bool isOverHighWatermark() const override { return over_high_watermark_; }
+  bool isReadable() const override { return isPeerShutDownWrite() || owned_buffer_.length() > 0; }
+
+private:
+  // Support isOpen() and close(). IoHandle owner must invoke close() to avoid potential resource
+  // leak.
+  bool closed_{false};
+
+  // The attached file event with this socket. The event is not owned by the socket.
+  Event::UserSpaceFileEventImpl* user_file_event_;
+
+  // Envoy never schedule two independent IO event on the same socket. Use this counter to verify.
+  int event_counter_{0};
+
+  // The schedulable handle of the above event.
+  Event::SchedulableCallbackPtr io_callback_;
+
+  // True if owned_buffer_ is not addable. Note that owned_buffer_ may have pending data to drain.
+  bool read_end_stream_{false};
+  Buffer::WatermarkBuffer owned_buffer_;
+
+  // Destination of the write().
+  WritablePeer* writable_peer_{nullptr};
```

Comment:
reference?

Reply:
Not viable right now. The high level picture is that we always have two `IoHandle` referencing each other but we must construct one after another.
If we can extract the `peer` out of `IoHandle` we can construct two peers and then construct two handles on top of the existing peers
However, it leaves two `peer`s referencing the two `IoHandle`s by pointer :(